crud/common/sharding/storage_metadata_cache.lua

```diff
@@ -0,0 +1,89 @@
+local stash = require('crud.common.stash')
+local utils = require('crud.common.sharding.utils')
+
+local storage_metadata_cache = {}
+
+local FUNC = 1
+local KEY = 2
+
+local cache_data = {
+    [FUNC] = nil,
+    [KEY] = nil,
+}
+
+local ddl_space = {
+    [FUNC] = '_ddl_sharding_func',
+    [KEY] = '_ddl_sharding_key',
+}
+
+local trigger_stash = stash.get(stash.name.ddl_triggers)
+
+local function update_sharding_func_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_func_def = utils.extract_sharding_func_def(tuple)
+    cache_data[FUNC][space_name] = utils.compute_hash(sharding_func_def)
+end
+
+local function update_sharding_key_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_key_def = tuple[utils.SPACE_SHARDING_KEY_FIELDNO]
+    cache_data[KEY][space_name] = utils.compute_hash(sharding_key_def)
+end
+
+local update_hash = {
+    [FUNC] = update_sharding_func_hash,
+    [KEY] = update_sharding_key_hash,
+}
+
+local function init_cache(section)
+    cache_data[section] = {}
+
+    local space = box.space[ddl_space[section]]
+
+    local update_hash_func = update_hash[section]
+
+    -- Remove old trigger if there was some code reload.
+    -- It is possible that ddl space was dropped and created again,
+    -- so removing non-existing trigger will cause fail;
+    -- thus we use pcall.
+    pcall(space.on_replace, space, nil, trigger_stash[section])
+
+    trigger_stash[section] = space:on_replace(
+        function(_, new)
+            return update_hash_func(new)
+        end
+    )
```

Comment:
Shouldn't it be no-op for the `'applier'` [session type](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/type/)?

Reply:
Cache may be useful for RO-instances too, so I think it shouldn't

Reply:
You're right.